nixops/backends/libvirtd.py

```diff
@@ -161,34 +147,64 @@ def iface(n):
-            "virsh",
-            "-c",
-            "qemu:///system",
-            "net-dhcp-leases",
-            "--network",
-            self.primary_net,
-        ]
-        lines = subprocess.check_output(cmd)
-        try:
-            i = lines.split().index(self.primary_mac)
-        except ValueError:
-            pass
+    def _fetch_ip(self):
+        if re.match("^\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}$", self.private_ipv4_setting):
+            return self.private_ipv4_setting
+
+        elif self.private_ipv4_setting == "arp":
+            # Inspired from https://rwmj.wordpress.com/2010/10/26/tip-find-the-ip-address-of-a-virtual-machine/
+            xml = subprocess.check_output(["virsh", "-c", "qemu:///system", "dumpxml", self.vm_id])
+            tree = ElementTree.fromstring(xml)
+            macs = [x.get("address") for x in tree.findall("devices/interface/mac")]
+            if len(macs) == 0:
+                raise Exception('VM has no interface configured; aborting')
+            self.log("Found MAC addresses " + repr(macs))
+
+            self.log_start("Waiting for IP address to appear in the ARP table...")
+            while True:
+                lines = subprocess.check_output(["arp", "-an"]).split("\n")
+                for line in lines:
+                    r = re.match('[^()]+ \(([0-9.]+)\) at ([0-9a-f:]+) ', line)
+                    if not r: continue
+                    for mac in macs:
+                        if r.group(2) == mac:
+                            ip = r.group(1)
+                            self.log_end(" " + ip)
+                            return ip
+
+                self.log_continue(".")
+                time.sleep(1)
+
+        elif self.private_ipv4_setting == "dhcp":
+            xml = subprocess.check_output(["virsh", "-c", "qemu:///system", "dumpxml", self.vm_id])
+            tree = ElementTree.fromstring(xml)
+            interfaces = tree.findall("devices/interface[@type='network']")
+            nets = [(x.find("source").get("network"), x.find("mac").get("address")) for x in interfaces]
+            if len(nets) == 0:
+                raise Exception('VM has no networks configured; aborting')
+            self.log("Found networks/MAC pairs " + repr(nets))
+
+            self.log_start("Waiting for IP address to appear in DHCP leases...")
+            while True:
+                for net, mac in nets:
+                    # TODO: parse command output with a proper regexp, and/or use "net-dhcp-leases --network <network> [--mac <mac>]"
+                    cmd = [ "virsh", "-c", "qemu:///system", "net-dhcp-leases", "--network", net ]
+                    lines = subprocess.check_output(cmd).split()
+                    try:
+                        i = lines.index(mac)
+                    except ValueError:
+                        continue
+                    else:
+                        ip_with_subnet = lines[i + 2]
+                        ip = ip_with_subnet.split('/')[0]
```

Comment:
Wait no, I had not looked at the context of your code. What you are doing is rather weird since it tells libvirt to give to the VM the IP it already has. Moreover, it has unreasonable assumptions like the fact that the given network has a dhcp server, which is exactly what I'm trying to get rid of